benches/fold_mut.rs

```diff
@@ -0,0 +1,109 @@
+use criterion::{black_box, criterion_group, criterion_main, Criterion};
+use itertools::Itertools;
+
+fn bench_sum(c: &mut Criterion) {
+    let mut group = c.benchmark_group("fold sum accumulator");
+
+    group.bench_function("fold", |b| {
+        b.iter(|| {
+            (0i64..1_000_000)
+                .map(black_box)
+                .fold(0, |sum, n| sum + n)
+        })
+    });
+
+    group.bench_function("fold_mut", |b| {
+        b.iter(|| {
+            (0i64..1_000_000).map(black_box).fold_mut(0, |sum, n| {
+                *sum += n;
+            })
+        })
+    });
+
+    group.finish();
+}
+
+fn bench_vec(c: &mut Criterion) {
+    let mut group = c.benchmark_group("fold vec accumulator");
+
+    group.bench_function("fold", |b| {
+        b.iter(|| {
+            (0i64..1_000_000)
+                .map(black_box)
+                .fold(Vec::new(), |mut v, n| {
+                    v.push(n);
+                    v
+                })
+        })
+    });
+
+    group.bench_function("fold_mut", |b| {
+        b.iter(|| {
+            (0i64..1_000_000)
+                .map(black_box)
+                .fold_mut(Vec::new(), |v, n| {
+                    v.push(n);
+                })
+        })
+    });
+
+    group.finish();
+}
+
+fn bench_num_with_chain(c: &mut Criterion) {
+    let mut group = c.benchmark_group("fold chained iterator with num accumulator");
+
+    group.bench_function("fold", |b| {
+        b.iter(|| {
+            (0i64..1_000_000)
+                .chain(0i64..1_000_000)
```

Comment:
Added a benchmark with `chain` to show that, because `fold_mut` uses `for_each`, it benefits from `Chain`'s specialization of [`fold`](https://doc.rust-lang.org/src/core/iter/adapters/chain.rs.html#107)